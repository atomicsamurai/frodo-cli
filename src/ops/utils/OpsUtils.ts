import { printError } from '../../utils/Console';

/**
 * Deep clone object
 * @param {any} obj object to deep clone
 * @returns {any} new object cloned from obj
 */
export function cloneDeep(obj) {
  return JSON.parse(JSON.stringify(obj));
}

/**
 * Performs an export or import given a function with its parameters with custom error handling that will just print the error if one is thrown and return null.
 * @param func The export or import function.
 * @param parameters The parameters to call the export or import function with. By default, it is { state }.
 * @returns {Promise<R | null>} Returns the result of the export or import function, or null if an error is thrown
 */
export async function exportOrImportWithErrorHandling<P extends any[], R>(
  func: (...params: P) => Promise<R>,
  parameters?: P
): Promise<R | null> {
  try {
    return await func(...parameters);
  } catch (error) {
    printError(error);
    return null;
  }
}
