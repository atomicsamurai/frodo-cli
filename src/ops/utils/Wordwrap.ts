export default function wordwrap(
  str: string,
  len: number,
  indent: string = undefined
) {
  if (!str) return '';
  return (
    str.match(
      new RegExp(
        `(\\S.{0,${len - 1 - (indent ? indent.length : 0)}})(?=\\s+|$)`,
        'g'
      )
    ) || []
  ).join(indent ? `\n${indent}` : '\n');
}
