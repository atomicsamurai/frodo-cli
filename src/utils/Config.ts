import { frodo, state } from '@rockcarver/frodo-lib';
import { AuthenticationSettingsSkeleton } from '@rockcarver/frodo-lib/types/api/AuthenticationSettingsApi';
import {
  FullExportInterface,
  FullGlobalExportInterface,
  FullRealmExportInterface,
} from '@rockcarver/frodo-lib/types/ops/ConfigOps';
import { ExportMetaData } from '@rockcarver/frodo-lib/types/ops/OpsTypes';
import fs from 'fs';
import os from 'os';

import { getIdmImportDataFromIdmDirectory } from '../ops/IdmOps';
import { getLegacyMappingsFromFiles } from '../ops/MappingOps';
import { getScriptExportByScriptFile } from '../ops/ScriptOps';
import { printMessage } from './Console';

const { getFilePath, readFiles } = frodo.utils;

const { exportFullConfiguration } = frodo.config;

const { getDefaultNoiseFilter } = frodo.cloud.log;

export const FRODO_CONFIG_PATH_KEY = 'FRODO_CONFIG_PATH';
export const FRODO_LOG_NOISEFILTER_FILENAME = 'LoggingNoiseFilter.json';

export function getConfigPath(): string {
  return process.env[FRODO_CONFIG_PATH_KEY] || `${os.homedir()}/.frodo`;
}

function getCustomNoiseFilters(): Array<string> {
  const filename = `${getConfigPath()}/${FRODO_LOG_NOISEFILTER_FILENAME}`;
  let noiseFilter = [];
  try {
    const data = fs.readFileSync(filename, 'utf8');
    noiseFilter = JSON.parse(data);
  } catch (e) {
    printMessage(`Error reading ${filename} (${e.message})`, 'error');
  }
  return noiseFilter;
}

export function getNoiseFilters(defaults: boolean): Array<string> {
  const filename = `${getConfigPath()}/${FRODO_LOG_NOISEFILTER_FILENAME}`;
  if (defaults) {
    printMessage(`Using default logging noise filters.`, 'info');
    return getDefaultNoiseFilter();
  }
  let noiseFilter = getCustomNoiseFilters();
  if (noiseFilter.length == 0) {
    printMessage(`No custom noise filters defined. Using defaults.`, 'info');
    noiseFilter = getDefaultNoiseFilter();
    try {
      fs.writeFileSync(filename, JSON.stringify(noiseFilter, null, 2));
      printMessage(
        `The default filters were saved in ${filename}. You can change the filters as needed.`,
        'info'
      );
    } catch (e) {
      printMessage(
        `Error creating noise filter configuration with default values.`,
        'error'
      );
    }
  }
  return noiseFilter;
}

/**
 * Gets the full export config from one of three locations:
 * 1. The file passed into the function if one is provided.
 * 2. The working directory if it exists (provided by the user)
 * 3. From deployment (e.g. cloud) if the exports are not locally provided
 * @param file The optional file path
 * @returns The full export config
 */
export async function getFullExportConfig(
  file: string | null = null
): Promise<FullExportInterface> {
  // Get export from file if it exists
  if (file) {
    return JSON.parse(fs.readFileSync(getFilePath(file), 'utf8'));
  }
  // If working directory doesn't exist, export from the cloud
  const workingDirectory = state.getDirectory();
  if (!workingDirectory) {
    return await exportFullConfiguration({
      useStringArrays: true,
      noDecode: false,
      coords: true,
      includeDefault: true,
      includeActiveValues: false,
      target: '',
    });
  }
  // Go through files in the working directory and reconstruct the full export
  return await getFullExportConfigFromDirectory(workingDirectory);
}

/**
 * Reconstructs the full export config from files in the given directory
 * @param directory The directory
 * @return The full export config
 */
export async function getFullExportConfigFromDirectory(
  directory: string
): Promise<FullExportInterface> {
  const realms = fs.readdirSync(directory + '/realm');
  const fullExportConfig: FullExportInterface = {
    meta: {} as ExportMetaData,
    global: {
      emailTemplate: {},
      idm: {},
      mapping: {},
      secrets: {},
      service: {},
      sync: {
        id: 'sync',
        mappings: [],
      },
      variables: {},
    },
    realm: Object.fromEntries(
      realms.map((r) => [
        r,
        {
          agents: {},
          application: {},
          authentication: {} as AuthenticationSettingsSkeleton,
          idp: {},
          managedApplication: {},
          policy: {},
          policyset: {},
          resourcetype: {},
          saml: {
            hosted: {},
            remote: {},
            metadata: {},
            cot: {},
          },
          script: {},
          service: {},
          theme: {},
          trees: {},
        },
      ])
    ),
  } as FullExportInterface;
  // Get global
  await getConfigFromDirectory(directory + '/global', fullExportConfig.global);
  // Get realms
  for (const realm of realms) {
    await getConfigFromDirectory(
      directory + '/realm/' + realm,
      fullExportConfig.realm[realm]
    );
  }
  return fullExportConfig;
}

async function getConfigFromDirectory(
  directory: string,
  exportConfig: FullGlobalExportInterface | FullRealmExportInterface
) {
  if (directory.startsWith('./')) {
    directory = directory.substring(2);
  }
  const samlPath = `${directory}/saml/`;
  const cotPath = `${directory}/cot/`;
  const idmPath = `${directory}/idm/`;
  const syncPath = `${directory}/sync/`;
  const scriptPath = `${directory}/script/`;

  const files = await readFiles(directory);
  const jsonFiles = files.filter((f) => f.path.endsWith('.json'));
  const samlFiles = jsonFiles.filter(
    (f) => f.path.startsWith(samlPath) || f.path.startsWith(cotPath)
  );
  const syncFiles = jsonFiles.filter((f) => f.path.startsWith(syncPath));
  const scriptFiles = jsonFiles.filter((f) => f.path.endsWith('.script.json'));
  const allOtherFiles = jsonFiles.filter(
    (f) =>
      !f.path.startsWith(samlPath) &&
      !f.path.startsWith(cotPath) &&
      !f.path.startsWith(idmPath) &&
      !f.path.startsWith(syncPath) &&
      !f.path.startsWith(scriptPath)
  );
  // Handle all other json files
  for (const f of allOtherFiles) {
    for (const [id, value] of Object.entries(
      JSON.parse(f.content) as Record<string, Record<string, object>>
    )) {
      if (id === 'meta') {
        continue;
      }
      if (exportConfig[id] == null) {
        exportConfig[id] = value;
      } else {
        Object.entries(value).forEach(([key, val]) => {
          exportConfig[id][key] = val;
        });
      }
    }
  }
  // Handle idm files
  if (fs.existsSync(idmPath)) {
    (exportConfig as FullGlobalExportInterface).idm =
      await getIdmImportDataFromIdmDirectory(idmPath);
  }
  // Handle sync files
  if (syncFiles.length) {
    (exportConfig as FullGlobalExportInterface).sync =
      await getLegacyMappingsFromFiles(syncFiles);
  }
  // Handle saml files
  for (const f of samlFiles) {
    let content = JSON.parse(f.content);
    content = content.saml;
    for (const [id, value] of Object.entries(
      content as Record<string, object>
    )) {
      if ((exportConfig as FullRealmExportInterface).saml[id] == null) {
        (exportConfig as FullRealmExportInterface).saml[id] = value;
      } else {
        Object.entries(value).forEach(
          ([key, val]) =>
            ((exportConfig as FullRealmExportInterface).saml[id][key] = val)
        );
      }
    }
  }
  // Handle extracted scripts
  for (const f of scriptFiles) {
    const scriptExport = getScriptExportByScriptFile(f.path);
    Object.entries(scriptExport.script).forEach(([id, script]) => {
      (exportConfig as FullRealmExportInterface).script[id] = script;
    });
  }
}

/**
 * Determines all locations where a string id is being used anywhere within the given configuration object
 * @param {object} configuration The configuration object
 * @param {string} id The id being searched for
 * @param {boolean} isEsv Whether the id corresponds to an ESV or not
 * @returns {string[]} an array of locations where the id is being used
 */
export function getIdLocations(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  configuration: object,
  id: string,
  isEsv: boolean
): string[] {
  return getIdLocationsRecurse(
    configuration,
    isEsv
      ? // For ESV ids, they contain either letters, numbers, dashes, or underscores. The dashes get replaced with periods (escaped with a \ for the regex)
        // since anywhere they are being used they will be used with periods, not dashes. Note that the (?:[^a-z0-9._]|$) expressions at the beginning and
        // end are meant to ensure that the id found is not a substring of some other id (i.e. the id found must either be at the beginning or end of the
        // string, or if in the middle of a string, is not preceded or followed by a character that would be part of another id).
        new RegExp(
          `(?:[^a-z0-9._]|^)${id.replaceAll('-', '\\.')}(?:[^a-z0-9._]|$)`
        )
      : // For normal ids, they contain only letters, numbers, or dashes.
        new RegExp(`(?:[^a-z0-9-]|^)${id}(?:[^a-z0-9-]|$)`)
  );
}

/**
 * Recursive helper for getIdLocations that finds locations of any strings contained in the configuration that pass the regex
 * @param {any} configuration The configuration (could be anything)
 * @param {RegExp} regex The regex test
 * @returns {string[]} an array of locations where the id is found
 */
function getIdLocationsRecurse(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  configuration: any,
  regex: RegExp
): string[] {
  let locations = [];
  const type = typeof configuration;
  if (type === 'object' && configuration !== null) {
    for (const [id, value] of Object.entries(
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      configuration as Record<string, any>
    )) {
      const usedLocations = getIdLocationsRecurse(value, regex);
      // Updates the relative locations of each place the id is used
      const updatedLocations = usedLocations.map(
        (loc) =>
          id +
          (value.name ? `(name: '${value.name}')` : '') +
          (loc === '' ? '' : '.') +
          loc
      );
      locations = locations.concat(updatedLocations);
    }
  }
  // This if statement determines whether or not the id is used in this configuration. If it is, return an empty string for its relative location.
  else if (type === 'string' && regex.test(configuration)) {
    locations.push('');
  }
  return locations;
}
