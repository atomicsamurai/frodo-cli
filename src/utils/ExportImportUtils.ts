import fs from 'fs';
import slugify from 'slugify';
import { ExportImportUtils } from '@rockcarver/frodo-lib';
import { printMessage } from './Console';

const { getMetadata } = ExportImportUtils;

export function getTypedFilename(name: string, type: string, suffix = 'json') {
  const slug = slugify(name.replace(/^http(s?):\/\//, ''));
  return `${slug}.${type}.${suffix}`;
}

/**
 * Save JSON object to file
 * @param {Object} data data object
 * @param {String} filename file name
 */
export function saveJsonToFile(data, filename) {
  const exportData = data;
  exportData.meta = getMetadata();
  fs.writeFile(filename, JSON.stringify(exportData, null, 2), (err) => {
    if (err) {
      return printMessage(`ERROR - can't save ${filename}`, 'error');
    }
    return '';
  });
}

export function saveToFile(type, data, identifier, filename) {
  const exportData = {};
  exportData['meta'] = getMetadata();
  exportData[type] = {};

  if (Array.isArray(data)) {
    data.forEach((element) => {
      exportData[type][element[identifier]] = element;
    });
  } else {
    exportData[type][data[identifier]] = data;
  }
  fs.writeFile(filename, JSON.stringify(exportData, null, 2), (err) => {
    if (err) {
      return printMessage(`ERROR - can't save ${type} to file`, 'error');
    }
    return '';
  });
}
