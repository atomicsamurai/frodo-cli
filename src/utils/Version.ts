import fs from 'fs';
import path from 'path';
import os from 'os';
import { fileURLToPath } from 'url';
import { compareVersions } from 'compare-versions';

import Color from 'colors';

import { LibVersion } from '@rockcarver/frodo-lib';

const VERSION_CACHE_FILE = `${os.homedir()}/.frodo/Versions.json`;
const VERSION_CHECK_INTERVAL = 30;

const GITHUB_REPOS_URL = `https://api.github.com/repos`;
const GITHUB_RELEASES_PATH_CLI = `/rockcarver/frodo-cli/releases`;
const GITHUB_RELEASES_PATH_LIB = `/rockcarver/frodo-lib/releases`;

const NPM_BASE_URL = `https://registry.npmjs.org`;
const NPM_PACKAGE_PATH_CLI = `/@rockcarver/frodo-cli`;
const NPM_PACKAGE_PATH_LIB = `/@rockcarver/frodo-lib`;

const ENDPOINTS = [
  {
    base: GITHUB_REPOS_URL,
    path: GITHUB_RELEASES_PATH_CLI,
  },
  {
    base: GITHUB_REPOS_URL,
    path: GITHUB_RELEASES_PATH_LIB,
  },
  {
    base: NPM_BASE_URL,
    path: NPM_PACKAGE_PATH_CLI,
  },
  {
    base: NPM_BASE_URL,
    path: NPM_PACKAGE_PATH_LIB,
  },
];

let versionObject = {
  last_checked: 0,
  cli: {
    github: '',
    npm: '',
  },
  lib: {
    github: '',
    npm: '',
  },
};

const __dirname = path.dirname(fileURLToPath(import.meta.url));

Color.enable();

const pkg = JSON.parse(
  fs.readFileSync(path.resolve(__dirname, '../../package.json'), 'utf8')
);

function getCliVersion() {
  return `${pkg.version}`;
}

function getLibVersion() {
  return LibVersion.getVersion();
}

function extractGithubReleaseInfo(data) {
  const release = {
    type: 'github',
    component: '',
    name: '',
    published_at: '',
  };
  // find first stable release
  const r = data.find((rel) => rel.prerelease == false);
  release.name = r.name;
  release.published_at = r.published_at;
  if (r.url.includes('frodo-cli')) {
    release.component = 'cli';
  } else {
    release.component = 'lib';
  }
  return release;
}

function extractNpmReleaseInfo(data) {
  const release = {
    type: 'npm',
    component: '',
    name: '',
    published_at: '',
  };
  // stable release
  release.name = data[`dist-tags`].latest;
  release.published_at = data.time[data[`dist-tags`].latest];
  if (data.name.includes('frodo-cli')) {
    release.component = 'cli';
  } else {
    release.component = 'lib';
  }
  return release;
}

async function getRemoteVersionData() {
  let useCache = true;
  if (fs.existsSync(VERSION_CACHE_FILE)) {
    const data = fs.readFileSync(VERSION_CACHE_FILE, 'utf8');
    versionObject = JSON.parse(data);
    if (
      versionObject.last_checked + VERSION_CHECK_INTERVAL <
      Math.floor(Date.now() / 1000)
    ) {
      useCache = false;
    }
  } else {
    useCache = false;
  }

  if (!useCache) {
    let allVersions = [];
    const result = await LibVersion.getAllVersions(ENDPOINTS);
    result.forEach((item) => {
      if (Array.isArray(item['value'].data)) {
        allVersions.push(extractGithubReleaseInfo(item['value'].data));
      } else {
        allVersions.push(extractNpmReleaseInfo(item['value'].data));
      }
    });
    // const allVersions = await LibVersion.getAllVersions(ENDPOINTS);
    allVersions.forEach((element) => {
      if (element.component == 'cli') {
        // cli
        if (element.type == 'github') {
          versionObject.cli.github = element.name;
        } else {
          versionObject.cli.npm = element.name;
        }
      } else {
        // lib
        if (element.type == 'github') {
          versionObject.lib.github = element.name;
        } else {
          versionObject.lib.npm = element.name;
        }
      }
    });
    versionObject.last_checked = Math.floor(Date.now() / 1000);
    fs.writeFileSync(
      VERSION_CACHE_FILE,
      JSON.stringify(versionObject, null, 2)
    );
  }
}

function getBinaryName() {
  return path.basename(process.execPath);
}

export async function getVersions(checkOnly: boolean) {
  let updateAvailable = false;
  let usingBinary = false;
  if (getBinaryName() == 'frodo') {
    usingBinary = true;
  }

  //   console.time('Execution Time');
  await getRemoteVersionData();
  //   console.timeEnd('Execution Time');
  let versionString = `You seem to be running the ${
    usingBinary ? 'binary' : 'NPM'
  } package`;

  versionString += `\nInstalled versions:`;
  versionString += `\ncli: v${getCliVersion()}\nlib: v${getLibVersion()}\nnode: ${
    process.version
  }`;
  let newVersionString = '';
  //   console.log(
  //     `${usingBinary}, ${compareVersions(getCliVersion(), versionObject.cli.npm)}`
  //   );
  if (
    (usingBinary &&
      compareVersions(getCliVersion(), versionObject.cli.github) == -1) ||
    (!usingBinary &&
      compareVersions(getCliVersion(), versionObject.cli.npm) == -1)
  ) {
    updateAvailable = true;
    newVersionString += `\n\nNewer version(s) available`;
    if (usingBinary) {
      newVersionString +=
        compareVersions(getCliVersion(), versionObject.cli.github) == -1
          ? `\ncli (github): v${versionObject.cli.github}`.brightGreen
          : ``;
    } else {
      newVersionString +=
        compareVersions(getCliVersion(), versionObject.cli.npm) == -1
          ? `\ncli (npm): v${versionObject.cli.npm}`.brightGreen
          : ``;
    }
  }
  if (checkOnly) {
    if (updateAvailable) {
      return `A new version of frodo is available.\nPlease run 'frodo -v' for more details.\n`
        .brightGreen;
    } else {
      return ``;
    }
  } else {
    return versionString + newVersionString;
  }
}
