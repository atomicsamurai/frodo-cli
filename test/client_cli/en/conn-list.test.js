import cp from 'child_process';
import { promisify } from 'util';
import { node14Compatibility } from '../utils/utils.js';

node14Compatibility();

const exec = promisify(cp.exec);
const CMD = 'frodo connections list --help';
const { stdout } = await exec(CMD);

test("CLI help interface for 'list' Usage should be expected english", async () => {
  // Arrange
  const expected = `
  Usage: frodo conn list [options]
    `.trim();
  // Act
  const testLine = stdout
    .split(/\n/)
    .find((line) => line.trim().startsWith('Usage:'))
    .trim();
  // Assert
  expect(testLine).toBe(expected);
});

test("CLI help interface 'connections List' description at line 2 should be expected english", async () => {
  // Arrange
  const expected = `
  List connection profiles.
    `.trim();
  // Act
  const testLine = stdout
    .split(/\n/)
    .map((line) => line.trim())
    .at(2);
  // Assert
  expect(testLine).toBe(expected);
});
