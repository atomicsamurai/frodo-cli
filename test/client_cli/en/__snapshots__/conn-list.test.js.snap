// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI help interface for 'conn list' should be expected english 1`] = `
"Usage: frodo conn list [options]

List connection profiles.

Options:
  -D, --directory <directory>          Set the working directory.
  --debug                              Debug output during command execution.
                                       If specified, may or may not produce
                                       additional output helpful for
                                       troubleshooting.
  --flush-cache                        Flush token cache.
  -h, --help                           Help
  --idm-host <idm-host>                IDM base URL, e.g.:
                                       https://cdk.idm.example.com/myidm. Use
                                       only if your IDM installation resides in
                                       a different domain and/or if the base
                                       path differs from the default
                                       "/openidm".
  -l, --long                           Long with all fields. (default: false)
  --login-client-id <client-id>        Specify a custom OAuth2 client id to use
                                       a your own oauth2 client for IDM API
                                       calls in deployments of type "cloud" or
                                       "forgeops". Your custom client must be
                                       configured as a public client and allow
                                       the authorization code grant using the
                                       "openid fr:idm:*" scope. Use the
                                       "--redirect-uri" parameter if you have
                                       configured a custom redirect uri
                                       (default:
                                       "<host>/platform/appAuthHelperRedirect.html").
  --login-redirect-uri <redirect-uri>  Specify a custom redirect URI to use
                                       with your custom OAuth2 client (efault:
                                       "<host>/platform/appAuthHelperRedirect.html").
  --no-cache                           Disable token cache for this operation.
  --sa-id <sa-id>                      Service account id.
  --sa-jwk-file <file>                 File containing the JSON Web Key (JWK)
                                       associated with the the service account.
  --verbose                            Verbose output during command execution.
                                       If specified, may or may not produce
                                       additional output.

Environment Variables:
  FRODO_HOST: AM base URL. Overridden by 'host' argument.
  FRODO_IDM_HOST: IDM base URL. Overridden by '--idm-host' option.
  FRODO_REALM: Realm. Overridden by 'realm' argument.
  FRODO_USERNAME: Username. Overridden by 'username' argument.
  FRODO_PASSWORD: Password. Overridden by 'password' argument.
  FRODO_LOGIN_CLIENT_ID: OAuth2 client id for IDM API calls. Overridden by '--login-client-id' option.
  FRODO_LOGIN_REDIRECT_URI: Redirect Uri for custom OAuth2 client id. Overridden by '--login-redirect-uri' option.
  FRODO_SA_ID: Service account uuid. Overridden by '--sa-id' option.
  FRODO_SA_JWK: Service account JWK. Overridden by '--sa-jwk-file' option but takes the actual JWK as a value, not a file name.
  FRODO_NO_CACHE: Disable token cache. Same as '--no-cache' option.
  FRODO_TOKEN_CACHE_PATH: Use this token cache file instead of '~/.frodo/TokenCache.json'.
  FRODO_CONNECTION_PROFILES_PATH: Use this connection profiles file instead of '~/.frodo/Connections.json'.
  FRODO_AUTHENTICATION_SERVICE: Name of a login journey to use.
  FRODO_DEBUG: Set to any value to enable debug output. Same as '--debug'.
  FRODO_MASTER_KEY_PATH: Use this master key file instead of '~/.frodo/masterkey.key' file.
  FRODO_MASTER_KEY: Use this master key instead of what's in '~/.frodo/masterkey.key'. Takes precedence over FRODO_MASTER_KEY_PATH.

"
`;
